### PR TITLE
New(workflow): Run Robot Framework tests with every push

### PR DESCRIPTION
**Pull request by rkubis**
## Description

This workflow builds `model-registry` image locally, starts it in `docker-compose` with `ml-metadata` and runs Robot Framework tests against it.

## How Has This Been Tested?

It successfully passed here in GitHub Actions.

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work.

### CODE REVIEW

File: .github/workflows/run-robot-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        uses: actions/setup-python@v4
+        with:
+          # Set Python version to install
+          python-version: '3.9'
+          # Set architecture of Python to install
+          architecture: 'x64'
+      # Install required Python packages for running Robot Framework tests
+      - name: Install required Python packages
+        # Install required Python packages using pip
+        run: pip install ml-metadata robotframework robotframework-requests pyyaml
```
Comment (lampajr) [inline .github/workflows/run-robot-tests.yaml]: May I suggest to export these libraries into a separated `requirements.txt` and then install it using `pip install -r test/robot/requirements.txt` ? In this way we can use the same command to install those libraries locally and furthermore I think it is much more maintainable.

File: .github/workflows/run-robot-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+name: run-robot-tests
+run-name: Run Robot Framework tests
+# Run workflow for every push to repository
+on: [push]
```
Comment (lampajr) [inline .github/workflows/run-robot-tests.yaml]: I think we should also add `pull_request` to run it on every PR too, wdyt?

**Review by lampajr — COMMENTED** [review]
Hey @rkubis, I took a quick look at this PR as I think this could be very helpful to run it every push/pr, thanks a lot!

I left some comments, let me know what you think :)

File: .github/workflows/run-robot-tests.yaml (hunk 2 above)
Comment (lampajr) [reply .github/workflows/run-robot-tests.yaml]: This way we can also check the workflow runs as expected in this PR

File: .github/workflows/run-robot-tests.yaml (hunk 1 above)
Comment (rkubis) [reply .github/workflows/run-robot-tests.yaml]: Thanks @lampajr, I agree!

File: .github/workflows/run-robot-tests.yaml (hunk 2 above)
Comment (rkubis) [reply .github/workflows/run-robot-tests.yaml]: @lampajr Sure, good idea :)

Comment (lampajr) [reply .github/workflows/run-robot-tests.yaml]: I was thinking something like https://github.com/opendatahub-io/model-registry/blob/259b39320953bf05942dcec1fb5ec74f7eb5d4a7/.github/workflows/build.yml#L3-L17

File: .github/workflows/run-robot-tests.yaml (hunk 1 above)
Comment (tarilabs) [reply .github/workflows/run-robot-tests.yaml]: +1 and kindly consider `ml-metadata==1.14.0`

Comment (rkubis) [reply .github/workflows/run-robot-tests.yaml]: @tarilabs Thanks, it is done now 👍

Comment (tonyxrmdavidson) [reply .github/workflows/run-robot-tests.yaml]: > +1 and kindly consider `ml-metadata==1.14.0`

Will there be a need to test different versions of ml-metadata?

File: .github/workflows/run-robot-tests.yaml (hunk 2 above)
Comment (rkubis) [reply .github/workflows/run-robot-tests.yaml]: @lampajr I used the same settings for `pull_request`, but for `push` I changed it from `main` to `*`.

**Review by tonyxrmdavidson — APPROVED** [review]
/lgtm

File: .github/workflows/run-robot-tests.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        # Install model_registry package as editable using pip
+        run: pip install -e clients/python
+      # Start docker compose with locally build image from current code
+      - name: Start docker compose with local image
+        # Start docker compose in the background
+        run: docker compose -f docker-compose-local.yaml up --detach
+      # Run Robot Framework tests against running docker compose
+      - name: Run Robot Framework tests
+        # Run Robot Framework tests from test/robot directory
+        run: robot test/robot
```
Comment (lampajr) [inline .github/workflows/run-robot-tests.yaml:57]: IIUC from https://github.com/opendatahub-io/model-registry/pull/158 to properly test all configs we should also add `TEST_MODE=Python robot test/robot/MRandLogicalModel.robot` in order to test the same `.robot` test using python library. Is my understading right @tarilabs ?

```suggestion
        run: | 
          robot test/robot
          TEST_MODE=Python robot test/robot/MRandLogicalModel.robot
```

**Review by lampajr — CHANGES_REQUESTED** [review]
Adding the suggestion below, tests should fail for:
```
[ ERROR ] Error in file '/home/alampare/repos/model-registry/test/robot/MRkeywords.resource' on line 5: Importing library '/home/alampare/repos/model-registry/test/robot/ModelRegistry.py' failed: AttributeError: module 'model_registry' has no attribute 'client'
Traceback (most recent call last):
  File "/home/alampare/repos/model-registry/test/robot/ModelRegistry.py", line 10, in <module>
    class ModelRegistry(mr.client.ModelRegistry):
PYTHONPATH:
  /home/alampare/repos/model-registry/test/robot
  /home/alampare/miniconda3/envs/model-registry/bin
  /home/alampare/miniconda3/envs/model-registry/lib/python39.zip
  /home/alampare/miniconda3/envs/model-registry/lib/python3.9
  /home/alampare/miniconda3/envs/model-registry/lib/python3.9/lib-dynload
  /home/alampare/miniconda3/envs/model-registry/lib/python3.9/site-packages
  /home/alampare/repos/model-registry/clients/python/src
```

As there was a change in the Python api [1], therefore the Robot tests should be adapated as well. This error is a good example of why adding this Robot test automation in every PR is a very good enhancement :)

[1] https://github.com/opendatahub-io/model-registry/pull/209/

File: .github/workflows/run-robot-tests.yaml (hunk 3 above)
Comment (tarilabs) [reply .github/workflows/run-robot-tests.yaml:57]: +1

**Review by antisaling — COMMENTED** [review]
Hey, Radim, thanks a lot for the changes :)

While I like the idea of having comments, I would also argue that a yaml file has a pretty simple syntax, and interleaving every line with comments is not really necessary.
It's also pretty common for comments to get outdated, or to be simply wrong, and this can hurt maintainability.
Either way I won't block you on this, it's just my 2c ;).

I've just submitted a fix for the failing imports, let's see if we can get this working (and passing!)

**Comment by antisaling** [comment]
I think the Python tests should pass cleanly now, can you rebase?

**Review by lampajr — APPROVED** [review]
lgtm for now, we might improve it in further iterations if needed.

As @isinyaaa raised I agree that most of the comments in the workflow could be removed but up to you @rkubis , not a blocker for me

**Comment by rkubis** [comment]
@isinyaaa Thanks for fixing Python tests for me! :)

**Comment by rkubis** [comment]
@isinyaaa @lampajr Yeah, I know that there is a lot more comments than expected 😄  I will leave it there for now and I can remove it in the next iteration. Thanks for feedback!

**Comment by tarilabs** [comment]
I notice this is approved and all tests a re passing.

@tonyxrmdavidson @rkubis just double check: I would assume you have merge rights to merge this when suitable; if you need someone to merge it on your behalf don't hesitate to ping/tag

**Comment by rkubis** [comment]
@tarilabs Thanks :) I don't have merge permission (yet), maybe @tonyxrmdavidson already have.

**Comment by lampajr** [comment]
Is there any update on this front? can we get this merged?

**Comment by rkubis** [comment]
@lampajr The update is that I got permission to merge this PR on my own right now 😄

**Pull request merged**